packages/style-engine/class-wp-style-engine-renderer.php

```diff
@@ -0,0 +1,193 @@
+		}
+
+		echo "<style>\n$output</style>\n";
+	}
+
+	/**
+	 * Filters incoming CSS properties against WordPress Core's allowed CSS attributes in wp-includes/kses.php.
+	 *
+	 * @param string $property_declaration A CSS property declaration, e.g., `color: 'pink'`.
+	 *
+	 * @return string A filtered CSS property. Empty if not allowed.
+	 */
+	public static function sanitize_property_declaration( $property_declaration ) {
+		return esc_html( safecss_filter_attr( $property_declaration ) );
+	}
+
+	/**
+	 * Creates a string consisting of CSS property declarations suitable for the value of an HTML element's style attribute.
+	 *
+	 * @param array $css_definitions An collection of CSS definitions `[ [ 'color' => 'red' ] ]`.
+	 *
+	 * @return string A concatenated string of CSS properties, e.g. `'color: red; font-size:12px'`
+	 */
+	public static function generate_inline_property_declarations( $css_definitions ) {
+		$css_rule_inline = '';
+
+		if ( empty( $css_definitions ) ) {
+			return $css_rule_inline;
+		}
+		foreach ( $css_definitions as $definition => $value ) {
+			$filtered_css = self::sanitize_property_declaration( "{$definition}: {$value}" );
+			if ( ! empty( $filtered_css ) ) {
+				$css_rule_inline .= $filtered_css . ';';
+			}
+		}
+		return $css_rule_inline;
+	}
+
+	/**
+	 * Creates a string consisting of a CSS rule.
+	 *
+	 * @param string $selector A CSS selector, e.g., `.some-class-name`.
+	 * @param array  $css_definitions An collection of CSS definitions `[ [ 'color' => 'red' ] ]`.
+	 * @param array  $options array(
+	 *      'prettify' => (boolean) Whether to add carriage returns and indenting.
+	 *      'indent' => (number) The number of tab indents to apply to the rule. Applies if `prettify` is `true`.
+	 *  );.
+	 *
+	 * @return string A CSS rule, e.g. `'.some-selector { color: red; font-size:12px }'`
+	 */
+	public static function generate_css_rule( $selector, $css_definitions, $options ) {
+		$css_rule_block = '';
+
+		if ( ! $selector || empty( $css_definitions ) ) {
+			return $css_rule_block;
+		}
+
+		$defaults       = array(
+			'prettify' => false,
+			'indent'   => 0,
+		);
+		$options        = wp_parse_args( $options, $defaults );
+		$indent         = str_repeat( "\t", $options['indent'] );
+		$css_rule_block = $options['prettify'] ? "$indent$selector {\n" : "$selector { ";
+
+		foreach ( $css_definitions as $definition => $value ) {
+			$filtered_css = self::sanitize_property_declaration( "{$definition}: {$value}" );
+			if ( ! empty( $filtered_css ) ) {
+				if ( $options['prettify'] ) {
+					$css_rule_block .= "\t$indent$filtered_css;\n";
+				} else {
+					$css_rule_block .= $filtered_css . ';';
+				}
+			}
+		}
+		$css_rule_block .= $options['prettify'] ? "$indent}\n" : ' }';
+		return $css_rule_block;
+	}
+
+	// @TODO Using cascade layers should be opt-in.
+	/**
+	 * Builds layers and styles rules from registered layers and styles for output.
+	 */
+	public static function enqueue_cascade_layers() {
```

Comment:
Just testing outputting cascade layers for now to test the concept. 

We really want registered styles to be the focus beforehand.

TODO:
- Needs to be tested against other styles for specificity
- How to make this opt-in? (We're instantiating `WP_Style_Engine` as a singleton, so should it be passed to the constructor somehow)